### PR TITLE
feat: main window size

### PR DESCRIPTION
**Pull request by KillerCodeMonkey**
- do not allow to resize main window
- scrollable time entry list
- bottom box with total hours
- label if no time entries found
- move function to format time entry notes for list to lib.ru + tests

### CODE REVIEW

**Comment by frenkel** [comment]
Works great. Sometimes the entries align exactly at the bottom, making it hard to see that you can scroll the area. How could we make this more obvious? I'm not sure how other programs handle this?

**Comment by KillerCodeMonkey** [comment]
dunno, but in general gtk applications are showing a thin dashed line, if there is something to scroll.

**Comment by KillerCodeMonkey** [comment]
one possible solution would be to always show the scrollbar, but this is a little bit ugly i think

**Comment by frenkel** [comment]
Thunderbird, Firefox and GNOME Terminal make the scrollbar always visible. In which application did you see a dashed line?

**Comment by KillerCodeMonkey** [comment]
Bildschirmfoto vom 2021-02-08 11-17-46
Do you see the light grey dashed line at the end? this thing is visible in the time-tracking app when there is something to scroll.

if you can scroll up it is on top, if down that at the bottom and when you can scroll up and down, it is on top and bottom.

Maybe it is a theme thing on my desktop?

**Comment by frenkel** [comment]
Seems like a feature of your distribution or theme indeed, mine (Fedora 33) doesn't show it:

Schermafdruk van 2021-02-08 12-05-50

**Comment by KillerCodeMonkey** [comment]
@frenkel i changed the scrollbar policies like described in the docs: https://gtk-rs.org/docs/gtk/trait.ScrolledWindowExt.html#tymethod.set_policy 
I changed it for the vertical scrollbar to Always, but then the scrollbar is only shown when the mouse hovers over the scrolled area.

So you can decide what you like more :).

**Comment by frenkel** [comment]
Thanks! Which one do you like best? I like this latest version.

**Comment by KillerCodeMonkey** [comment]
i do not have any special fav. so feel free to merge :)

**Comment by frenkel** [comment]
Is it possible to scroll the view to the bottom when you add a new entry? They now appear outside of the view if you had scrolled upwards previously.

**Comment by KillerCodeMonkey** [comment]
i searched for something like this in the ScrolledWindowExt, but i did not find anything. :S

**Comment by KillerCodeMonkey** [comment]
@frenkel i implemented the scrolling, but sadly it seems like that the new entry is not in the list (rendered?) when i try to scroll. It is scrolling, but just to the previous last entry. 
Maybe you see the problem?

**Comment by frenkel** [comment]
Thanks! I’ll try to take a look next weekend!

File: src/ui.rs [hunk 1]
```diff
@@ -350,6 +351,16 @@ impl Ui {
         self.set_total(total_hours);
 
         self.grid.show_all();
+
+        match scroll_bottom {
+            Some(scroll) => {
+                if scroll {
+                    let adjustment = self.scroll_view.get_vadjustment().unwrap();
+                    adjustment.set_value(adjustment.get_upper() - adjustment.get_page_size());
```
Comment (frenkel) [inline src/ui.rs]: It appears this code works from within a size adjustment event for grid. Something like this works:

```
self.grid.connect_size_allocate(clone!(scroll_view => move |_,_| {
  let adj = scroll_view.get_vadjustment().unwrap();
  adj.set_value(adj.get_upper() - adj.get_page_size());
}));
```
However, after adding this event handler, any other event will still scroll to bottom, resulting in an unusable edit of the first item for example. Any ideas?

File: src/ui.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ macro_rules! clone {
 
 pub enum Signal {
     SetTitle(String),
-    SetTimeEntries(Vec<TimeEntry>),
+    SetTimeEntries(Vec<TimeEntry>, Option<bool>),
```
Comment (frenkel) [inline src/ui.rs]: Why not use just `bool` instead of `Option<bool>`?

File: src/app.rs [hunk 3]
```diff
@@ -120,7 +120,7 @@ impl App {
             .expect("Sending message to ui thread");
     }
 
-    fn retrieve_time_entries(&mut self) {
+    fn retrieve_time_entries(&mut self, scroll_bottom: Option<bool>) {
```
Comment (frenkel) [inline src/app.rs]: Why not use just `bool` instead of `Option<bool>`?

File: src/ui.rs [hunk 4]
```diff
@@ -250,7 +251,7 @@ impl Ui {
         self.total_amount_label.set_label(&formatted_label);
     }
 
-    pub fn set_time_entries(&mut self, time_entries: Vec<TimeEntry>) {
+    pub fn set_time_entries(&mut self, time_entries: Vec<TimeEntry>, scroll_bottom: Option<bool>) {
```
Comment (frenkel) [inline src/ui.rs]: Why not use just `bool` instead of `Option<bool>`?

**Review by frenkel — CHANGES_REQUESTED** [review]
I found the solution, but bumped into another problem, see my comment. Can you get it to work correctly?

File: src/ui.rs (hunk 1 above)
Comment (KillerCodeMonkey) [reply src/ui.rs]: thats why i am only try to scroll if creating a new item without using this event handler.

dumb idea from my side: add such a handler on top of the "set_time_entries" method, if a new item was added (scroll_bottom param is true) and at the end we could try to remove this listener.

I will check it

**Comment by frenkel** [comment]
Your “dumb idea” might work as the only solution 😀

**Comment by KillerCodeMonkey** [comment]
but i am really bad at searching through such docs. do you know how to remove an event listener?

**Comment by KillerCodeMonkey** [comment]
ah maybe just `disconnect(handlerId)`

**Comment by frenkel** [comment]
I think you would need to use something based on `signal::signal_handler_disconnect()`.
Another solution might be to introduce a new `Signal` called "ScrollToBottom" which you would send after the RetrieveTimeEntries signal from the new timer part. That would be more elegant and would call the scroll code just a tiny bit later, which might solve your original bug.

**Comment by KillerCodeMonkey** [comment]
after hours of testing and trying... nothing worked. i was not able (maybe because of the lack of rust knowledge) to get it work with the event listener.

first i tried to just globally listen on grid size change and have a scroll_bottom propery on Ui. But i was not able to read the current value of `self.scroll_buttom` inside the closure.

second tried to add a listener per call of `set_time_entries` and remove it inside of the closure... but no luck either.

Then i thought i am ultra clever -> just add a while loop and check, if all children of the grid are `realized`, but it seems everything is realized directly after the `grid.show_all`, which leads to the same strange behavior that it is only scrolled to the second last element.

Seems like everything is updated, but somehow the changing of the dimensions is done later.

So i ended up just adding a little timeout... dumb, but it seems to work.

**Comment by frenkel** [comment]
That's too bad, thanks for your testing and investigation! I'll soon have a look myself to try and avoid the timeout.

**Comment by frenkel** [comment]
This is a tough one. I couldn't get it to work correctly either. I do want to avoid the timeout though, slow computers might need even more timeout, that will only give trouble. I believe we need to setup the size-allocate signal once and somehow signal (via the global state in UI?) that we do want to scroll or not.

**Comment by KillerCodeMonkey** [comment]
I Tried IT already, but i was not able to Pass The boolean (If to scroll or
Not) into The closure :(.

I guess The timeout will Work in every Machine.

If i Understand IT correctly, after The show_all call everything, but The
size allocation is done.


Am 14.02.2021 in 17:42, Frank Groeneveld <notifications@github.com>
schrieb:

This is a tough one. I couldn't get it to work correctly either. I do want
to avoid the timeout though, slow computers might need even more timeout,
that will only give trouble. I believe we need to setup the size-allocate
signal once and somehow signal (via the global state in UI?) that we do
want to scroll or not.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/frenkel/timer-for-harvest/pull/32#issuecomment-778803573>,
or unsubscribe

**Comment by frenkel** [comment]
Last night I tried a number of different things:

- Look in the size-allocate signal at the adjustment and see if you can retrieve the old adjustment. If they differ, we want to scroll. Unfortunately the old adjustment doesn't seem to be available anywhere.
- Look in the changed signal of vadjustment of the scroll_view. This should fire if anything but the value of the vadjustment changes (see https://gtk-rs.org/docs/gtk/trait.AdjustmentExt.html#tymethod.connect_changed) but it appears to fire when you scroll manually as well.

That last option seems to be the nicest way, if we can get it to work correctly. As I understand it, value is your current scroll position. If the total size of the scroll area changes, upper of vadjustment changes and it should fire the signal.

**Comment by KillerCodeMonkey** [comment]
so maybe we can store the latest height and if the new height > the one before -> scroll and update the latest height. But there i do not really know again, how i can move the property to the closure of the signal callback to read and write it.

**Comment by frenkel** [comment]
Yes, that gets quite complex, using an Arc: https://doc.rust-lang.org/std/sync/struct.Arc.html

**Comment by frenkel** [comment]
I've just pushed a working implementation. I'll test this for a few days before merging and releasing a new version.

**Comment by KillerCodeMonkey** [comment]
One Thing that Cane to My mind Last night:

Do we Need to handle/avoid scrolling, If the height is Changing when switching days?

**Comment by frenkel** [comment]
I thought about that as well. We could probably set the new current height variable to the lower value on switch. I’ll see how it works first the coming days. We’re almost there 😀

**Comment by frenkel** [comment]
I'm building a new release now, thank you very much for this great contribution! :tada:

**Comment by KillerCodeMonkey** [comment]
YEAAAAHHAAAA!

🚀

Am Fr., 26. Feb. 2021 um 09:22 Uhr schrieb Frank Groeneveld <
notifications@github.com>:

**Comment by KillerCodeMonkey** [comment]
@frenkel if i try to install the latest version i get:

require libglib2.0-0 (>= 2.64.6) but 2.64.3-1~ubuntu20.04.1 was installed.
Seems like the requested dependency version does not exist for ubuntu 20.04

**Comment by frenkel** [comment]
Same issue as https://github.com/frenkel/timer-for-harvest/issues/33? Try updating first.

**Comment by KillerCodeMonkey** [comment]
i tried, but there are no updates for that lib... strange

**Comment by KillerCodeMonkey** [comment]
i got it. thought it is part of a stable update, but i have to enable preview updates

**Comment by KillerCodeMonkey** [comment]
There it is. the beautiful "minimal" scrollable main window :D

**Pull request merged**